cmd/kubeadm/app/apis/kubeadm/validation/validation_test.go

```diff
@@ -458,3 +475,158 @@ func TestValidateFeatureGates(t *testing.T) {
+				DiscoveryTokenUnsafeSkipCAVerification: false,
+				DiscoveryTokenAPIServers:               []string{"192.168.122.100:9898"},
+			},
+			true,
+		},
+		{
+			"WARNING: kubeadm doesn't fully support multiple API Servers yet",
+			&kubeadm.NodeConfiguration{
+				DiscoveryToken:           "abcdef.1234567890123456@foobar",
+				DiscoveryTokenAPIServers: []string{"192.168.122.100:9898", "192.168.122.88:9898"},
+			},
+			true,
+		},
+		{
+			"valid: DiscoveryFile with DiscoveryTokenAPIServers",
+			&kubeadm.NodeConfiguration{
+				DiscoveryFile:            "https://url/file.conf",
+				DiscoveryTokenAPIServers: []string{"192.168.122.100:9898"},
+			},
+			true,
+		},
+		{
+			"valid: DiscoveryFile without DiscoveryTokenAPIServers",
+			&kubeadm.NodeConfiguration{
+				DiscoveryFile: "https://url/file.conf",
+			},
+			true,
+		},
+		{
+			"valid: using token-based discovery with DiscoveryTokenCACertHashes",
+			&kubeadm.NodeConfiguration{
+				DiscoveryToken:                         "abcdef.1234567890123456@foobar",
+				DiscoveryTokenAPIServers:               []string{"192.168.122.100:9898"},
+				DiscoveryTokenCACertHashes:             []string{"sha256:7173b809ca12ec5dee4506cd86be934c4596dd234ee82c0662eac04a8c2c71dc"},
+				DiscoveryTokenUnsafeSkipCAVerification: false,
+			},
+			true,
+		},
+		{
+			"valid: using token-based discovery with DiscoveryTokenCACertHashe but skip ca verification",
+			&kubeadm.NodeConfiguration{
+				DiscoveryToken:                         "abcdef.1234567890123456@foobar",
+				DiscoveryTokenAPIServers:               []string{"192.168.122.100:9898"},
+				DiscoveryTokenCACertHashes:             []string{"sha256:7173b809ca12ec5dee4506cd86be934c4596dd234ee82c0662eac04a8c2c71dc"},
+				DiscoveryTokenUnsafeSkipCAVerification: true,
+			},
+			true,
+		},
+	}
+	for _, rt := range tests {
+		err := ValidateArgSelection(rt.c, nil).ToAggregate()
+		if (err == nil) != rt.expected {
+			t.Errorf(
+				"%s test case failed: ValidateArgSelection:\n\texpected: %t\n\t  actual: %t",
+				rt.name,
+				rt.expected,
+				(err == nil),
+			)
+		}
+	}
+}
+
+func TestValidateJoinDiscoveryTokenAPIServer(t *testing.T) {
+	var tests = []struct {
+		s        *kubeadm.NodeConfiguration
+		expected bool
+	}{
+		{
+			&kubeadm.NodeConfiguration{
+				DiscoveryTokenAPIServers: []string{"192.168.122.100"},
+			},
+			false,
+		},
+		{
+			&kubeadm.NodeConfiguration{
+				DiscoveryTokenAPIServers: []string{"192.168.122.100:9898"},
+			},
+			true,
+		},
+	}
+	for _, rt := range tests {
+		actual := ValidateJoinDiscoveryTokenAPIServer(rt.s, nil)
+		if (len(actual) == 0) != rt.expected {
+			t.Errorf(
+				"failed ValidateJoinDiscoveryTokenAPIServer:\n\texpected: %t\n\t  actual: %t",
+				rt.expected,
+				(len(actual) == 0),
+			)
+		}
+	}
+}
+
+func TestValidateDiscoveryFile(t *testing.T) {
+	var tests = []struct {
+		s        string
+		expected bool
+	}{
+		{"foo", false},
+		{"http://[::1]a", false},
```

Comment:
can you add a couple of more cases here?

Reply:
done.